### PR TITLE
Option parsing

### PR DESCRIPTION
**Pull request by wthrowe**
I have made many changes from the original code.  Some of the highlights:
* Most of the documentation and error output has been rewritten, including more checks for incorrect input.
* The parser-enforced value checks have been simplified: bounds checks on individual container elements are gone, fixed size vectors are gone (but could still be obtained using min size=max size), and all remaining constraints are shown in the help output.
* Formatting checks are done on option help.  (These existed before but were not run due to a bug.)
* `MAKE_CREATABLE_FROM_YAML` macro for non-factory stuff as discussed elsewhere.
* A "forwarding" form for `MAKE_CREATABLE_FROM_YAML` to allow parsing simple types without using full option parsing (e.g., parsing an old-style Point class as `Center: [1, 2, 3]` or an enum as `Norm: L2`).
* Constructors called from Factory and MCFY now take an `OptionContext` parameter that can be passed to the `PARSE_ERROR` macro to include information about the input file location in an error message (generally for doing more complex checks on input values).
* `OptionsList` in factory/MCFY classes has been renamed to `options` because it should be camel case and it isn't a list in MCFY forwarding.
* Added `Options<...>::apply

I've put some examples of parsing error output at http://astro.cornell.edu/~wthrowe/parsing_errors (executable at http://astro.cornell.edu/~wthrowe/parsing_errors.cpp).

### CODE REVIEW

**Comment by wthrowe** [comment]
Builds are failing because yaml-cpp is too old.  The latest release can't handle non-copyable types.  We need something more recent than November 2016.

**Comment by nilsdeppe** [comment]
Okay, I can update the container we run on Travis

**Comment by wthrowe** [comment]
I've prepared a commit documenting the yaml-cpp version requirement, but I'll hold off pushing until it can trigger a Travis run with the update.

File: src/Options/Options.hpp [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+
+template <typename OptionsList>
+template <typename T,
+          std::enable_if_t<Options_details::has_lower_bound<T>::value>*>
+inline void Options<OptionsList>::check_lower_bound(
+    const typename T::type& t, const OptionContext& context) const {
+  static_assert(cpp17::is_same_v<decltype(T::lower_bound()), typename T::type>,
+                "Lower bound is not of the same type as the option.");
+  static_assert(not cpp17::is_same_v<typename T::type, bool>,
+                "Cannot set an lower bound for a bool.");
```
Comment (kidder) [inline src/Options/Options.hpp]: an -> a

File: src/Options/Options.hpp [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/StdHelpers.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \ingroup OptionParsing
+/// Information about the nested operations being performed by the
+/// parser, for use in printing errors.  A default-constructed
+/// OptionContext is printed as an empty string.  This struct is
+/// primarily used as an argument to PARSE_ERROR for reporting input
+/// file parsing errors.
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: Could you add a brief explanation of what `YAML::Mark` is used for or link to the yaml-cpp documentation? The yaml-cpp documentation is not exactly super complete or easy to navigate. If I understand correctly, `YAML::Mark` used to store the line and column in the file where a parsing error occurred?

File: src/Utilities/ConstantExpressions.hpp [hunk 3]
```diff
@@ -168,13 +168,13 @@ inline constexpr auto make_array_from_list() {
 
 /// \ingroup ConstantExpressions
 /// \brief Compute the length of a const char* at compile time
-SPECTRE_ALWAYS_INLINE constexpr size_t cstring_length(const char* str) {
```
Comment (nilsdeppe) [inline src/Utilities/ConstantExpressions.hpp]: Can the place that caused the error use a `std::string` instead which actually supports these features?

File: src/Options/Options.hpp [hunk 4]
```diff
@@ -0,0 +1,557 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defined classes and functions that handle parsing of input parameters.
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: "Defines"

File: src/Options/Options.hpp [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+/// primarily used as an argument to PARSE_ERROR for reporting input
+/// file parsing errors.
+struct OptionContext {
+  bool top_level{true};
+  std::string context;
+  YAML::Mark mark{YAML::Mark::null_mark()};
+
+  void append(const std::string& c) { context += c + ":\n"; }
+};
+inline std::ostream& operator<<(std::ostream& s, const OptionContext& c) {
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: blank line

File: src/Options/Options.hpp [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+  return s << c.context << Options_details::mark_info(c.mark);
+}
+
+/// \ingroup OptionParsing
+/// The type that options are passed around as.  Contains YAML node
+/// data and an OptionContext.
+class Option_t {
+ public:
+  Option_t() = delete;
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: add comment saying we don't want a default constructor

File: src/Options/Options.hpp [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+
+  /// NOTE: This constructor overwrites the mark data in the supplied
+  /// context with the one from the node.
+  Option_t(YAML::Node node, OptionContext context = OptionContext())
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: explicit

File: src/Options/Options.hpp [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+      : node_(std::move(node)), context_(std::move(context)) {
+    context_.mark = node_.Mark();
+  }
+
+  Option_t(OptionContext context) : context_(std::move(context)) {}
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: explicit

File: src/Options/Options.hpp [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+
+  const YAML::Node& Node() const { return node_; }
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: noexcept

Comment (nilsdeppe) [inline src/Options/Options.hpp]: lower case first letter

File: src/Options/Options.hpp [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+}
+
+/// \ingroup OptionParsing
+/// \brief Class that handles parsing an input file
+///
+/// \details Options is templated on OptionsList, which must be a
+/// typelist of structs.  Each of these structs corresponds to a
+/// parameter to be read from an input file.  These structs have the
+/// following form:
+/// \code
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: Could we use snippets from the tests to replace this code block?

File: src/Options/Options.hpp [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+    PARSE_ERROR(error_context, e.message());
+  } catch (std::exception& e) {
+    ERROR("Unexpected exception: " << e.what());
+  }
+}
+
+/// \ingroup OptionParsing
+/// \brief Class that handles parsing an input file
+///
+/// \details Options is templated on OptionsList, which must be a
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `OptionsList`

File: src/Options/Options.hpp [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+template <typename OptionsList>
+class Options {
+ public:
+  /// \param help_text an overall description of the options
+  explicit Options(std::string help_text);
+
+  /// Parse a string to obtain options and their values.
+  ///
+  /// \param options the string holding the YAML formatted options
+  void parse(const std::string& options);
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: blank line

File: src/Options/Options.hpp [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+                                 const OptionContext& /*context*/) const {}
+  //@}
+
+  //@{
+  /// Check for an iterable type that the size is above the upper bound
+  template <typename T,
+            std::enable_if_t<
+                Options_details::has_upper_bound_on_size<T>::value>* = nullptr>
+  void check_upper_bound_on_size(const typename T::type& t,
+                                   const OptionContext& context) const;
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: indentation

File: src/Options/Options.hpp [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+                T>::value>* = nullptr>
+  void check_upper_bound_on_size(const typename T::type& /*t*/,
+                                 const OptionContext& /*context*/) const {}
+  //@}
+
+  //@{
+  /// Returns the default value or errors if there is no default.
+  template <typename T,
+            std::enable_if_t<Options_details::has_default<T>::value>* = nullptr>
+  inline typename T::type get_default() const {
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: remove redundant inlines

File: src/Options/Options.hpp [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+  /// Check for an iterable type that the size is above the upper bound
+  template <typename T,
+            std::enable_if_t<
+                Options_details::has_upper_bound_on_size<T>::value>* = nullptr>
+  void check_upper_bound_on_size(const typename T::type& t,
+                                   const OptionContext& context) const;
+  template <typename T,
+            std::enable_if_t<not Options_details::has_upper_bound_on_size<
+                T>::value>* = nullptr>
+  void check_upper_bound_on_size(const typename T::type& /*t*/,
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `constexpr` empty functions

File: src/Options/Options.hpp [hunk 16]
```diff
@@ -0,0 +1,557 @@
...
+}
+
+template <typename OptionsList>
+void Options<OptionsList>::parse(const YAML::Node& options) {
+  if (not(options.IsMap() or options.IsNull())) {
+    PARSE_ERROR(context_,
+                "'" << options << "' does not look like options.\n"
+                << help());
+  }
+  for (auto it = options.begin(); it != options.end(); ++it) {
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: range-based for loop

File: src/Options/Options.hpp [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+                  << parsing_help(options));
+    }
+
+    // Check for duplicate key
+    if (0 != parsed_options_.count(key)) {
+      PARSE_ERROR(context,
+                  "Key '" << key << "' specified twice.\n"
+                  << parsing_help(options));
+    }
+    parsed_options_.emplace(std::make_pair(key, value));
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: don't need `make_pair`

File: src/Options/Options.hpp [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+struct apply_helper;
+
+template <typename... Tags>
+struct apply_helper<tmpl::list<Tags...>> {
+  template <typename Options, typename F>
+  static decltype(auto) apply(const Options& opts, F&& func) {
+    return func(opts.template get<Tags>()...);
+  }
+};
+}
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `// namespace Options_details`

File: src/Options/Options.hpp [hunk 19]
```diff
@@ -0,0 +1,557 @@
...
+
+template <typename OptionsList>
+template <typename T,
+          std::enable_if_t<Options_details::has_lower_bound_on_size<T>::value>*>
+void Options<OptionsList>::check_lower_bound_on_size(
+    const typename T::type& t, const OptionContext& context) const {
+  static_assert(cpp17::is_same_v<decltype(T::lower_bound_on_size()), size_t>,
+                "lower_bound_on_size() is not a size_t.");
+  const auto size = t.end() - t.begin();
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: use `t.size()` if possible

File: src/Options/Options.hpp [hunk 20]
```diff
@@ -0,0 +1,557 @@
...
+}
+
+template <typename OptionsList>
+template <typename T,
+          std::enable_if_t<Options_details::has_upper_bound_on_size<T>::value>*>
+void Options<OptionsList>::check_upper_bound_on_size(
+    const typename T::type& t, const OptionContext& context) const {
+  static_assert(cpp17::is_same_v<decltype(T::upper_bound_on_size()), size_t>,
+                "upper_bound_on_size() is not a size_t.");
+  const auto size = t.end() - t.begin();
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `t.size()`

File: src/Options/Options.hpp [hunk 21]
```diff
@@ -0,0 +1,557 @@
...
+  if (static_cast<size_t>(size) > T::upper_bound_on_size()) {
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: Then remove static_cast

File: src/Options/Options.hpp [hunk 22]
```diff
@@ -0,0 +1,557 @@
...
+    if (not convert<std::map<K, V>>::decode(node, ordered)) {
+      return false;
+    }
+    rhs.clear();
+    rhs.insert(std::make_move_iterator(ordered.begin()),
+               std::make_move_iterator(ordered.end()));
+    return true;
+  }
+};
+}
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `// namespace YAML`

File: src/Options/OptionsDetails.hpp [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// The string used in option structs
+using OptionString_t = const char* const;
+
+/// Holds details of the implementation of Options
+namespace Options_details {
+template <typename S, typename U = void>
```
Comment (nilsdeppe) [inline src/Options/OptionsDetails.hpp]: `U = void` -> `= cpp17::void_t<>`

File: src/Options/OptionsDetails.hpp [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+  template <typename T,
+            std::enable_if_t<not has_upper_bound_on_size<T>::value>* = nullptr>
+  static std::string print_upper_bound_on_size() {
+    return "";
+  }
+
+  template <typename T>
+  void operator()(tmpl::type_<T> /*meta*/) {
+    std::ostringstream ss;
+    ss << "  " << std::setw(max_label_size_) << std::left << T::label
```
Comment (nilsdeppe) [inline src/Options/OptionsDetails.hpp]: look at `setw` things

File: src/Options/OptionsDetails.hpp [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+      limits += limit;
+    }
+    if (not limits.empty()) {
+      ss << " [" << limits << "]";
+    }
+    ss << "\n" << std::setw(max_label_size_ + 2) << ""
+       << std::setw(0) << T::help << "\n\n";
+    value += ss.str();
+  }
+  value_type value{};
```
Comment (nilsdeppe) [inline src/Options/OptionsDetails.hpp]: blank line above

File: src/Options/OptionsDetails.hpp [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+  if (m.is_null()) {
+    return "";
+  } else {
+    std::ostringstream ss;
+    ss << "At line " << m.line + 1 << " column " << m.column + 1 << ":\n";
+    return ss.str();
+  }
+}
+
+template <typename T>
```
Comment (nilsdeppe) [inline src/Options/OptionsDetails.hpp]: move to `TypeTraits.hpp`

File: tests/Unit/Options/Test_Options.cpp [hunk 27]
```diff
@@ -0,0 +1,502 @@
...
+struct DefaultedOption {
+  using type = int;
+  static constexpr OptionString_t label = {"Defaulted"};
+  static constexpr OptionString_t help = {"Option with a default value"};
+  static type default_value() { return 3; }
+};
+}  // namespace
+
+TEST_CASE("Unit.Options.Defaulted.specified", "[Unit][Options]") {
+  const std::string yaml = "Defaulted: 20";
```
Comment (nilsdeppe) [inline tests/Unit/Options/Test_Options.cpp]: inline short strings

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Options/Factory.hpp [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+#include "Utilities/StdHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup OptionParsing
+/// \brief Allows creating a pointer to an abstract base class from an
+/// input file.
+///
+/// To use a factory, the base class (here `BaseClass`) should inherit
+/// from Factory<BaseClass> and define a static method \code
+///  static std::string class_id() { return "BaseClass"; }
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: Could experiment with using `pretty_type` to replace `class_id()`

File: src/Options/Factory.hpp [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+/// \endcode
+/// and a type \code
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `\code` to next line

File: src/Options/Factory.hpp [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+#include "Options/Options.hpp"
+#include "Utilities/StdHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup OptionParsing
+/// \brief Allows creating a pointer to an abstract base class from an
+/// input file.
+///
+/// To use a factory, the base class (here `BaseClass`) should inherit
+/// from Factory<BaseClass> and define a static method \code
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `\code` on next line

File: src/Options/Factory.hpp [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+///  using creatable_classes = tmpl::list<Derived1, ...>;
+/// \endcode
+///
+/// Derived classes should:
+/// 1) define a static member function std::string class_id() which
+///    returns a unique id for the derived class, similar to the base
+///    class method shown above.
+/// 2) define static OptionString_t help containing class-specfic help
+///    text
+/// 3) define a type options as a typelist of option structs required
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: define a type alias `options`

File: src/Options/Factory.hpp [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+  using Base_t = BaseClass;
+
+  /// Create a derived object.
+  static std::unique_ptr<BaseClass> create(const Option_t& options);
+
+ private:
+  static std::string help_derived();
+
+  template <
+      typename CreateLs,
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `CreateLs` to `CreateList`

File: src/Options/Factory.hpp [hunk 33]
```diff
@@ -0,0 +1,181 @@
...
+      typename std::enable_if_t<(tmpl::size<CreateLs>::value != 0)>* = nullptr>
+  static std::unique_ptr<BaseClass> create_derived(const std::string& id,
+                                                   const Option_t& opts);
+
+  template <
+      typename CreateLs,
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `CreateLs`

File: src/Options/Factory.hpp [hunk 34]
```diff
@@ -0,0 +1,181 @@
...
+      typename std::enable_if_t<(tmpl::size<CreateLs>::value == 0)>* = nullptr>
+  static std::unique_ptr<BaseClass> create_derived(const std::string& /*id*/,
+                                                   const Option_t& /*opts*/) {
+    return std::unique_ptr<BaseClass>{};
+  }
+};
+
+template <typename BaseClass>
+template <typename CreateLs,
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `CreateLs`

File: src/Options/Factory.hpp [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+    derived_opts.set_node(node.begin()->second);
+  } else {
+    PARSE_ERROR(options.Context(),
+                "Expected a class or a class with options, got:\n"
+                << node);
+  }
+  derived_opts.Context().append("While creating type " + id);
+  auto derived =
+      create_derived<typename BaseClass::creatable_classes>(id, derived_opts);
+  if (derived) {
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `derived != nullptr` for clarity

File: src/Options/Factory.hpp [hunk 36]
```diff
@@ -0,0 +1,181 @@
...
+    const size_t name_col = 2;
+    const size_t help_col = 22;
+    const size_t end_col = 80;
+
+    std::ostringstream ss;
+    ss << std::left
+       << std::setw(name_col) << ""
+       << std::setw(help_col - name_col - 1) << T::class_id();
+    if (ss.str().size() >= help_col) {
+      ss << std::setw(0) << "\n" << std::setw(help_col - 1) << "";
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: Don't need to `setw(0)`

File: src/Options/Factory.hpp [hunk 37]
```diff
@@ -0,0 +1,181 @@
...
+struct convert<std::unique_ptr<T>> {
+  static bool decode(const Node& node, std::unique_ptr<T>& rhs) {
+    Option_t options(node);
+    options.Context().append("While creating a " + T::class_id());
+    options.Context().top_level = false;
+    rhs = T::create(options);
+    return true;
+  }
+};
+}
```
Comment (nilsdeppe) [inline src/Options/Factory.hpp]: `// namespace YAML`

File: tests/Unit/Options/Test_Factory.cpp [hunk 38]
```diff
@@ -0,0 +1,218 @@
...
+  TestWithArg(std::string arg, const OptionContext& /*context*/)
+      : arg_(std::move(arg)) {}
+
+  std::string name() const override { return class_id() + "(" + arg_ + ")"; }
+
+ private:
+  std::string arg_;
+};
+
+struct OptionsVecArg {
```
Comment (nilsdeppe) [inline tests/Unit/Options/Test_Factory.cpp]: clean up

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by wthrowe** [comment]
The new commits fall into a few groups:
1. Minor fixes, largely based on above comments
2. Removing class_id and option labels in favor of using the name of the class/struct.  Assuming we don't run into any issues with compiler/library support, I think the class_id changes are a clear win.  The option label changes sometimes force you to name option structs in ways that may conflict with other names.  This can be avoided by wrapping options in an additional struct, so I think we're still OK, but the change is not completely downside free.
3. Redesigning MAKE_CREATABLE_FROM_YAML, since in trying to write an example I found some serious problems with the previous version.  (This does mean that `T::options` is always a typelist again, so it can be changed back to `T::options_list` if anyone cares.)

There was discussion of adding support for giving the option parser a small set of values allowed for an option.  There is currently support for adding parsability to an enum (see example on MCFY) and for providing parsing errors from factory class constructors.  The only additional feature that I think the parser could add would be displaying options in the help text, but we're trying to keep that short so I'm kind of leery of printing arbitrary lists of values there.

**Comment by wthrowe** [comment]
And looks like yaml-cpp is broken on the clang Travis setups.

**Comment by wthrowe** [comment]
Trying to clean up the doxygen warnings, I've hit two issues:
1. It looks like doxygen doesn't understand `decltype(auto)` return type.
2. One of the corrected `\snippet` lines is over 80 characters, and doxygen doesn't allow the line to be broken.  I guess we just have to use short tag names in files with long names?

File: src/Options/Options.hpp [hunk 39]
```diff
@@ -32,89 +32,120 @@
...
   /// Sets the node and updates the context's mark to correspond to it.
-  void set_node(YAML::Node node) {
-    node_ = std::move(node);
+  void set_node(YAML::Node node) noexcept {
+    // clang-tidy: YAML::Node not movable (as of yaml-cpp-0.5.3)
+    node_ = std::move(node);  // NOLINT
     context_.mark = node_.Mark();
   }
 
+  /// Convert to an object ot type `T`.
```
Comment (kidder) [inline src/Options/Options.hpp]: ot -> of

File: src/Options/Options.hpp [hunk 40]
```diff
@@ -356,37 +381,37 @@ void Options<OptionsList>::parse(const std::string& options) {
...
+  if (not(node.IsMap() or node.IsNull())) {
     PARSE_ERROR(context_,
-                "'" << options << "' does not look like options.\n"
+                "'" << node << "' does not look like options.\n"
                 << help());
   }
-  for (auto it = options.begin(); it != options.end(); ++it) {
-    const auto key = it->first.as<std::string>();
-    const auto value = it->second;
+  for (const auto& kv : node) {
```
Comment (kidder) [inline src/Options/Options.hpp]: while I've used kv for the range-based key-value pair, others  (  @gsb76 , @osheamonn ) have voiced other opinions

File: src/Utilities/PrettyType.hpp [hunk 41]
```diff
@@ -684,4 +684,21 @@ template <typename T>
 std::string get_runtime_type_name(const T& x) {
   return boost::core::demangle(typeid(x).name());
 }
+
+/*!
+ * \ingroup PrettyType
+ * \brief Extract the "base name" from a name, that is, the name
+ * without template parameters or scopes.
+ */
+std::string extract_base_name(std::string name);
```
Comment (kidder) [inline src/Utilities/PrettyType.hpp]: I cannot think of an alternative at the moment, but when I first saw base_name I thought about the name of a base class...

File: src/Options/Factory.hpp [hunk 42]
```diff
@@ -26,24 +27,17 @@
...
 /// \code
 ///  using creatable_classes = tmpl::list<Derived1, ...>;
 /// \endcode
 ///
 /// Derived classes should:
-/// 1) define a static member function std::string class_id() which
-///    returns a unique id for the derived class, similar to the base
-///    class method shown above.
-/// 2) define static OptionString_t help containing class-specfic help
+/// 1) define static OptionString_t help containing class-specfic help
```
Comment (kidder) [inline src/Options/Factory.hpp]: doxygen does have a syntax for [enumerated list](https://www.stack.nl/~dimitri/doxygen/manual/lists.html) (having to relabel the comments made me think of this...)

File: src/Options/MakeCreatableFromYaml.hpp [hunk 43]
```diff
@@ -12,50 +12,52 @@
...
-    typename T,
-    std::enable_if_t<tt::is_a_v<tmpl::list, typename T::options>>* = nullptr>
-T create(const Option_t& option) {
+/// \ingroup OptionParsing
+/// Used by MAKE_CREATABLE_FROM_YAML to create an object.  The default
+/// action is to parse options using `T::options`.  This function may
+/// be specialized to change that behavior for specific types.
+///
+/// Do not call this function directly.  Use Option_t::parse_as
+/// instead.
```
Comment (kidder) [inline src/Options/MakeCreatableFromYaml.hpp]: merge to previous line?

File: src/Options/MakeCreatableFromYaml.hpp [hunk 44]
```diff
@@ -12,50 +12,52 @@
...
 ///
-/// The "option" form allows the class to be parsed from a set of
-/// key-value pairs.  The class must define a typelist `options` of
-/// structs appropriate for passing to class Options, an
-/// OptionString_t `help`, a default constructor, a move assignment
-/// operator, and a constructor taking the arguments corresponding to
-/// the `options` and an OptionContext.
+/// A class to be created must be default constructible and move
+/// assignable.  If create_from_yaml is not specialized for the class
+/// it must additionally declare a tnpl::list of Options-style option
```
Comment (kidder) [inline src/Options/MakeCreatableFromYaml.hpp]: tnpl -> tmpl

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by wthrowe** [comment]
Speculation on the Travis yaml-cpp problem: using a libstdc++-linked yaml-cpp for a libc++ build.

**Comment by nilsdeppe** [comment]
Yes, that is what it is.

File: src/Options/Options.hpp [hunk 45]
```diff
@@ -0,0 +1,582 @@
...
+
+  /// Parse a string to obtain options and their values.
+  ///
+  /// \param options the string holding the YAML formatted options
+  void parse(const std::string& options) noexcept;
+
+  /// Parse an Option_t to obtain options and their values.
+  void parse(const Option_t& options);
+
+  /// Parse an file containing options
```
Comment (kidder) [inline src/Options/Options.hpp]: Parse a

File: src/Options/Options.hpp [hunk 46]
```diff
@@ -0,0 +1,582 @@
...
+  template <typename T,
+            std::enable_if_t<not Options_details::has_lower_bound_on_size<
+                T>::value>* = nullptr>
+  constexpr void check_lower_bound_on_size(
+      const typename T::type& /*t*/,
+      const OptionContext& /*context*/) const noexcept {}
+  //@}
+
+  //@{
+  /// Check for an iterable type that the size is above the upper bound
```
Comment (kidder) [inline src/Options/Options.hpp]: above -> below

File: src/Options/Options.hpp [hunk 47]
```diff
@@ -0,0 +1,582 @@
...
+                << pretty_type::short_name<T>() << "'.\n"
+                << help());
+  }
+  //@}
+
+  //@{
+  /// If the options has a lower bound, check it is satisfied.
+  ///
+  /// Note: Lower bounds are >=, not just >.
+  /// \tparam T the option "type"
```
Comment (kidder) [inline src/Options/Options.hpp]: "type" -> something clearer

File: src/Options/Options.hpp [hunk 48]
```diff
@@ -0,0 +1,582 @@
...
+  constexpr void check_lower_bound(
+      const typename T::type& /*t*/,
+      const OptionContext& /*context*/) const noexcept {}
+  //@}
+
+  //@{
+  /// If the options has a upper bound, check it is satisfied.
+  ///
+  /// Note: Upper bounds are <=, not just <.
+  /// \tparam T the option "type"
```
Comment (kidder) [inline src/Options/Options.hpp]: ditto

File: src/Options/Options.hpp [hunk 49]
```diff
@@ -0,0 +1,582 @@
...
+  template <typename Options, typename F>
+  static decltype(auto) apply(const Options& opts, F&& func) {
+    return func(opts.template get<Tags>()...);
+  }
+};
+}  // namespace Options_details
+
+// \cond
+// Doxygen is confused by decltype(auto)
+template <typename OptionsList>
```
Comment (kidder) [inline src/Options/Options.hpp]: OptionsList -> OptionList

File: tests/Unit/Options/Test_Options.cpp [hunk 50]
```diff
@@ -0,0 +1,409 @@
...
+  static constexpr OptionString_t help = {
+    "Option with bounds and a default value"};
+  // These are optional
+  static type default_value() { return 3; }
+  static type lower_bound() { return 2; }
+  static type upper_bound() { return 10; }
+};
+/// [options_example_scalar_struct]
+}  // namespace
+
```
Comment (kidder) [inline tests/Unit/Options/Test_Options.cpp:96]: test that a out-of-bounds default fails

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Options/Options.hpp [hunk 51]
```diff
@@ -232,7 +232,10 @@ class Options {
   void parse(const YAML::Node& node);
 
   //@{
-  /// Check for an iterable type that the size is above the lower bound
+  /// Check for an that the size is not smaller than the lower bound
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: `Check that the....`

File: src/Options/Options.hpp [hunk 52]
```diff
@@ -247,7 +250,10 @@ class Options {
   //@}
 
   //@{
-  /// Check for an iterable type that the size is below the upper bound
+  /// Check for an that the size is not larger than the upper bound
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: again

File: src/Options/Options.hpp [hunk 53]
```diff
@@ -263,6 +269,8 @@ class Options {
 
   //@{
   /// Returns the default value or errors if there is no default.
+  ///
+  /// \tparam T the option struct
   template <typename T,
             std::enable_if_t<Options_details::has_default<T>::value>* = nullptr>
```
Comment (nilsdeppe) [inline src/Options/Options.hpp]: Change all to `Requires<...> = nullptr`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
squash

**Comment by kidder** [comment]
Will, can you rebase on the latest commit that attempted to fix Travis stuff

**Comment by nilsdeppe** [comment]
I'll look at what the heck Travis is doing

**Comment by nilsdeppe** [comment]
Could you rebase to get the new Travis things that uses libstdc++ with Clang?

**Comment by nilsdeppe** [comment]
@wthrowe could you try applying the changes in this commit https://github.com/nilsdeppe/spectre/commit/0dfcc1563353219b7debcc59a31dd3d905249ece ?

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**